sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldTypeDescriptors.java

```diff
@@ -0,0 +1,119 @@
+ */
+package org.apache.beam.sdk.schemas;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.ImmutableBiMap;
+import java.lang.reflect.ParameterizedType;
+import java.util.Collection;
+import java.util.Map;
+import org.apache.beam.sdk.schemas.Schema.FieldType;
+import org.apache.beam.sdk.schemas.Schema.TypeName;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.sdk.values.TypeDescriptors;
+import org.joda.time.Instant;
+/**
+ * Utilities for converting between {@link Schema} field types and {@link TypeDescriptor}s that
+ * define Java objects which can represent these field types.
+ */
+public class FieldTypeDescriptors {
+  private static final BiMap<TypeName, TypeDescriptor> PRIMITIVE_MAPPING =
+      ImmutableBiMap.<TypeName, TypeDescriptor>builder()
+          .put(TypeName.BYTE, TypeDescriptors.bytes())
+          .put(TypeName.INT16, TypeDescriptors.shorts())
+          .put(TypeName.INT32, TypeDescriptors.integers())
+          .put(TypeName.INT64, TypeDescriptors.longs())
+          .put(TypeName.DECIMAL, TypeDescriptors.bigdecimals())
+          .put(TypeName.FLOAT, TypeDescriptors.floats())
+          .put(TypeName.DOUBLE, TypeDescriptors.doubles())
+          .put(TypeName.STRING, TypeDescriptors.strings())
+          .put(TypeName.DATETIME, TypeDescriptor.of(Instant.class))
+          .put(TypeName.BOOLEAN, TypeDescriptors.booleans())
+          .put(TypeName.BYTES, TypeDescriptor.of(byte[].class))
+          .build();
+  /** Get a {@link TypeDescriptor} from a {@link FieldType}. */
+  public static TypeDescriptor javaTypeForFieldType(FieldType fieldType) {
+    switch (fieldType.getTypeName()) {
+      case ARRAY:
+        return TypeDescriptors.lists(javaTypeForFieldType(fieldType.getCollectionElementType()));
+      case MAP:
+        return TypeDescriptors.maps(
+            javaTypeForFieldType(fieldType.getMapKeyType()),
+            javaTypeForFieldType(fieldType.getMapValueType()));
+      case ROW:
+        return TypeDescriptors.rows();
+      default:
+        return PRIMITIVE_MAPPING.get(fieldType.getTypeName());
+    }
+  }
+  /** Get a {@link FieldType} from a {@link TypeDescriptor}. */
+  public static FieldType fieldTypeForJavaType(TypeDescriptor typeDescriptor) {
+    if (typeDescriptor.isArray()
+        || typeDescriptor.isSubtypeOf(TypeDescriptor.of(Collection.class))) {
+      return getArrayFieldType(typeDescriptor);
+    } else if (typeDescriptor.isSubtypeOf(TypeDescriptor.of(Map.class))) {
+      return getMapFieldType(typeDescriptor);
+    } else if (typeDescriptor.isSubtypeOf(TypeDescriptor.of(Row.class))) {
+      throw new IllegalArgumentException(
+          "Cannot automatically determine a field type from a Row class"
+              + " as we cannot determine the schema. You should set a field type explicitly.");
```

Comment:
Just an idea, would it be possible (and practical) to generate row subclasses for all schemas, so that you can access the schema in places like this one? E.g.


```
class GeneratedRowSubclassWithSchema$1 extends Row {
   public static final Schema ROW_SCHEMA = ...;
}

if (typeDescriptor.isSubtypeOf(TypeDescriptor.of(Row.class)) {
   // do reflection magic to extract ROW_SCHEMA static field
}
```